### ISSUE TITLE
Missing validation options

### ISSUE THREAD

**Issue by norpan** [NoRedInk/json-elm-schema#6]
I have tried to collect the currently missing validation options. Will implement them and make a PR if that's ok. Some of them will require breaking changes, so let's discuss them.

- [ ] `multipleOf` (`NumberSchema`)
  _Just an additional `Maybe num` field_
- [ ] `items` with array of schemas and `additionalItems` (`ArraySchema`)
  _Make a union type and incorporate `additionalItems` to the array case._
- [ ] `exclusiveMaximum` and `exclusiveMinimum` (`NumberSchema`)
  _Add extra `Bool` parameter to `minItems` and `maxItems` and make `exclusiveMaximum` and `exclusiveMinimum` construction functions with the same type signature as `minimum` and `maximum`_
- [ ] `uniqueItems` (`ArraySchema`)
  _Just an additional `Bool` field (default `False`)_
- [ ] `maxProperties` and `minProperties` (`ObjectSchema`)
 _Just two `Maybe Int`fields_
- [ ] `patternProperties` (`ObjectSchema`)
  _`Dict String Schema` or `List`?_
- [ ] `additionalProperties` (`ObjectSchema`)
- [ ] `dependencies` (`ObjectSchema`)
  _`Dict String Schema or `List`?_
- [ ] `not` (`BaseSchema`)
  _Just an additional `Maybe Schema` field_
- [ ] `definitions` (`BaseSchema`)
- [ ] `default` (`BaseSchema`)
- [ ] `id` (`BaseSchema`)
- [ ] `$schema` (`BaseSchema`)
- [ ] `$ref`
  _(make `Schema` a union type, or inline `$ref`s)_

**Comment by jwoudenberg** [comment]
Awesome list, thanks for putting that together! Some thoughts below. Let me know what you think!

---

I think it might make sense to create a separate `TupleSchema` constructor for the alternative use of `items`. Apart from not breaking backwards compatibility, this has the advantage of allow us to make `minItems` and `maxItems` only available for the `ArraySchema` and `additionalItems` only available for the `TupleSchema`.

With regards to `exclusiveMinimum` and `exclusiveMaximum`: does it make sense to have both a normal limit and an inclusive limit set at the same time? If not, maybe we keep a single `minimum` and `maximum` property on the model and assign it a `type Limit = Inclusive Int | Exclusive Int`. I think this doesn't even have to break backwards compatibility if we ensure the `minimum` and `maximum` property-setter functions keep setting the inclusive version as before (we could add separate `exclusive` versions of these).

Should we wrap `uniqueItems` in a `Maybe` too, just to ensure a JSON schema generated from it is as 'clean' as possible?

Never heard of `dependencies` before, cool property! Can we store these inside of the `Property` they depend on? That would prevent us from accidentally creating dependencies for non-existing properties.

With regards to `definitions`, I wonder if we need this. Because we're working with Elm, we can create reusable subschema's simply by assigning a schema to a variable and using that variable wherever we need it in the main schema. Two downsides to that approach:
- It does not allow for resursive subschema's
- If the resulting schema contains the same reused chunk inlined in multiple places it might become large and unreadable.
Perhaps a middle way is to create an `inlineRef : String -> Schema -> Schema` function, which keeps track of which Schema's should be extracted upon encoding to an actual Json schema. Then we only pull them out into a definitions block at the last possible moment (upon encoding).

**Comment by norpan** [comment]
> I think it might make sense to create a separate TupleSchema constructor for the alternative use of items. Apart from not breaking backwards compatibility, this has the advantage of allow us to make minItems and maxItems only available for the ArraySchema and additionalItems only available for the TupleSchema.

`maxItems` are available for the `TupleSchema` if `additionalItems` is true (or the empty object, which is the default). But yes, this can be encoded easily, so having `TupleSchema` is good, although the name can be misleading.

> With regards to exclusiveMinimum and exclusiveMaximum: does it make sense to have both a normal limit and an inclusive limit set at the same time? If not, maybe we keep a single minimum and maximum property on the model and assign it a type Limit = Inclusive Int | Exclusive Int. I think this doesn't even have to break backwards compatibility if we ensure the minimum and maximum property-setter functions keep setting the inclusive version as before (we could add separate exclusive versions of these).

Yes, this is what I meant. Compatibility with the Schema type alias is broken though (a field is not just added, but changed).

> Should we wrap uniqueItems in a Maybe too, just to ensure a JSON schema generated from it is as 'clean' as possible?

Not sure, it's debatable. Should we use Maybe or substitute default values? It's simple enough to check if a value is the default and not output it.

> Never heard of dependencies before, cool property! Can we store these inside of the Property they depend on? That would prevent us from accidentally creating dependencies for non-existing properties.

Yes, that's a good idea. An example of this is the schema for JSON Schema itself, it has:
```
    "dependencies": {
        "exclusiveMaximum": [ "maximum" ],
        "exclusiveMinimum": [ "minimum" ]
    },
```

> With regards to definitions, I wonder if we need this. Because we're working with Elm, we can create reusable subschema's simply by assigning a schema to a variable and using that variable wherever we need it in the main schema. 

Yes, we don't need it for recursion (recursion is not allowed with $refs, they are inlined verbatim), but it's present in JSON schemas that you would like to read in, and can reference external schemas. So it makes sense to store them as is, I think.

If you want to be able to create them, then using `inlineRef` is a good option I think.

**Comment by jwoudenberg** [comment]
> maxItems are available for the TupleSchema if additionalItems is true (or the empty object, which is the default). But yes, this can be encoded easily, so having TupleSchema is good, although the name can be misleading.

👍 

> Yes, this is what I meant. Compatibility with the Schema type alias is broken though (a field is not just added, but changed).

True, but I because that particular alias is not exported outside the library, any code written with this library before this change should still be valid afterwards. I think Elm package is actually smart enough to pick up on that and give us a minor version bump because of it.

> Not sure, it's debatable. Should we use Maybe or substitute default values? It's simple enough to check if a value is the default and not output it.

That's definitely true. Also it might make a cleaner API, allowing you to just pass in `uniqueItems` (without any arguments) to set that particular flag.

**Comment by norpan** [comment]
> True, but I because that particular alias is not exported outside the library, any code written with this library before this change should still be valid afterwards.
Yes, I was sort of hoping to be able to use the actual schema to create Forms, but  that functionality could be added to the package too. And a validator.

I'm going to make the easy ones first though :)

**Comment by jwoudenberg** [comment]
@norpan Oh noes, I pressed 'comment' before finishing my reply. Finished up my answer above :).

**Comment by norpan** [comment]
> I think it might be possible to read them into the inlined format while decoding. When we encounter an internal ref during decoding, we can look for the corresponding definition (could be on definitions or somewhere else on the JSON structure), decode that into a Schema, pass it to inlineRef and put that into the structure.
External refs would need a separate approach (perhaps ref : String -> Schema ?), which I think makes sense. The internal ones we can ensure are correct, the external ones not.

Actually I was wrong, recursion is allowed, but only in children, you are not allowed to recursively `allOf` for instance.

I think we need to allow recursion though. It's pretty useful. And since we don't have (and don't want) circular structures in Elm, we need to keep the actual string keys. An example is the Json Schema schema again, which uses "#" as reference to a subschema.

http://json-schema.org/schema

**Comment by jwoudenberg** [comment]
We can probably make recursing work by having `inlineRef` take a lambda. This is the way recursive structures are supported in decoders and html:

`inlineRef "my-schema" (\_ -> object [])`

**Comment by norpan** [comment]
Hmmm, I don't understand how that would work. Say I want to write the http://json-schema.org/schema, how would I write it?

**Comment by jwoudenberg** [comment]
The way I see it, definitions in json schema's are basically variables. In Elm, we have those too, so I'd probably prefer to use those to define the small schema components:

```elm
-- Some small subschemas
schemaArray : Schema
schemaArray = ...

positiveInteger : Schema
positiveInteger = ...

-- The top level schema can reference all these subschema's.
schema : Schema
schema = ...
```

This fails when you try to do recursion, as you rightly pointed out. To get around this, a bunch of libs include a `lazy` function to work around this (see for instance [Json.Decode](http://package.elm-lang.org/packages/elm-lang/core/latest/Json-Decode). Perhaps our `inlineRef` is really just another example of that. With that, a recursive structure might be expressed as this:

```elm
tree : Schema
tree =
    object
        [ properties
            [ required "value" (string [])
            , required "children" <|
                array
                    [ items (lazy (\_ -> tree))
                    ]
            ]
        ]
```

This would require an extra sum constuctor: `Schema = ... | Lazy (a -> Schema)`.

**Comment by norpan** [comment]
Yeah, I'm not asking you to write the whole thing, although that would be a good test case :)

Using a lambda takes care of the problem with infinite evaluation of the constructor function, but the problem with doing it the way you suggest is that we are putting pieces together locally, but the `$ref` to something would vary depending on where in the schema you put it. If you'd take your `tree` defined above and use it as a subschema, the `$ref` you have created in it would point to the wrong place (it would be `#`).

The decoder case is actually easier, the decoder has access to the whole schema and can check that references exists globally during decoding, and even inline them. I'm just not sure how to make those constructor functions work.

**Comment by jwoudenberg** [comment]
I'm not sure how to make them work either, but I'm convinced it can be done and kind of itching to give it a try :). Do you mind if I pick up implementation of that particular part of the spec?

**Comment by norpan** [comment]
Please do! I'm going to do the easy bits now, and continue building on my validator/form generator. :)
